### PR TITLE
data-api: Continue ProcessResourceDefinitions through un-marshaling various schema/definition information 

### PR DESCRIPTION
**Pull request by mbfrahry**
There is still quite a few todos to knock out as some data is missing but this provides the scaffold to quickly add that information when it is readily available. 

Would it be better to figure out where that information is hiding and fix it in this PR or in a new PR?

### CODE REVIEW

File: tools/data-api/internal/repositories/generator_models.go
Comment (tombuildsstuff) [inline tools/data-api/internal/repositories/generator_models.go:48]: these are both Optional therefore need to be defined as pointers:

```suggestion
	RequestObject                    *ObjectDefinition `json:"RequestObject,omitempty"`
	ResponseObject                   *ObjectDefinition `json:"ResponseObject,omitempty"`
```

File: tools/data-api/internal/repositories/generator_models.go
Comment (tombuildsstuff) [inline tools/data-api/internal/repositories/generator_models.go:46]: Resource Ids are optional (an operation can be just a Uri Suffix [e.g. `/some-action`]), so this needs to become a pointer:

```suggestion
	ResourceId                       *string           `json:"ResourceId,omitempty"`
```

File: tools/data-api/internal/repositories/generator_models.go
Comment (tombuildsstuff) [inline tools/data-api/internal/repositories/generator_models.go:63]: This needs to become an Object Definition - see #3135

File: tools/data-api/internal/repositories/generator_models.go
Comment (tombuildsstuff) [inline tools/data-api/internal/repositories/generator_models.go:22]: This wants renaming - since that needs to be done in the other package - let's sort that package, then rely on it here.

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
Leaving a partial review - but most of these need to be fixed in the other package (since this one is based on that) - as such I'll take a deeper dive into that package now, once that's fixed up we should be able to reuse it here

**Comment by tombuildsstuff** [comment]
#3223 updates the Model/Field/ObjectDefinition types relevant for this PR - but I guess it'd probably be best to hold off with this one until all of the models are refactored/reviewed, so that we can duplicate all of the models at once?

File: tools/data-api/internal/repositories/services.go
Comment (tombuildsstuff) [inline tools/data-api/internal/repositories/services.go:270]: this one should be sourceable from `field.ContainsDiscriminatedTypeValue`?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @stephybun 

I've taken a look through and left a number of comments inline for sake of completeness - but on the whole this is looking pretty good. The three main themes for comments are:

1. Ordering the deserialization, so that when mapping the ObjectDefinition (etc) we can validate that the Constant X is exposed when it's marked as a Reference. That ensures we're handling any invalid data both when saving it (in the importers for generated code) and when parsing the data (which handles handwritten code). This means we need to load Constants then Models then Resource IDs and then finally Operations.
2. Adding mappings between the models used for persistence and those we're returning from the `repositories` package - since the backing values for constants aren't necessarily 1:1 (as the types aren't, since the Terraform Tests will likely be output into `*.hcl` files in the future, etc)
3. Updating the `processX` methods to be focused on parsing the file, then returning a mapping function - this means we could unit test the critical parts in the future if we wanted too more easily.

However on the whole this is looking pretty good, if we can fix those up then this otherwise LGTM 👍 

Thanks!

File: tools/data-api/internal/repositories/services.go
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:171]: I think it's quite implicit that we're dealing with the API items/things here, are we trying to differentiate version and resource from other versions or resources within the data API? If not I actually find it confusing to append these with `api` since it makes me question what other possible types of version and resource there are.

File: tools/data-api/internal/repositories/services.go
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:34]: > Given there's going to be multiple sub-directories of API Definitions within the ./api-definitions directory - it'd be worth supporting that sooner rather than later.

The services repository directory is constructed out of the root dir for the api definitions and a service type via which we route by so as far as I'm concerned this is already supported?

> Presumably we'll want to identify any sub-directories containing a metadata.json file in this method, so that we can validate that each of the Services are unique

Two points here:

1. Given there's already a mechanism for determining which service type to read the api definitions in for I don't currently see a value in identifying sub-dirs that contain a `metadata.json`, I guess at most to validate that the `Source` value in the `metadata.json` corresponds to the api definitions we're wanting to load.
2. Checking that there aren't multiple definitions for the same service makes sense

I think 2. this is a good isolated small task that can and in the interest of time be completed in a separate PR. I've left a TODO in the method.

File: tools/data-api/internal/repositories/services.go
Comment (tombuildsstuff) [reply tools/data-api/internal/repositories/services.go:171]: TBH this was mostly in reference to `resource` which is ambiguous between the Resource within an API Version and a Terraform Resource - `apiVersion` was mostly for consistency, since we'll likely end up reusing this in other functions.

Providing we stick with the terminology `resource` and `terraformResource` to differentiate those two that'd work fine 👍

File: tools/data-api/internal/repositories/services.go
Comment (tombuildsstuff) [reply tools/data-api/internal/repositories/services.go:34]: For 1:

The relationship between `ServiceType` and `Source` within the `metadata.json` isn't (necessarily) 1:1 - we're going to need to port over [the existing hand-defined data](https://github.com/hashicorp/pandora/tree/main/data/Pandora.Definitions.HandDefined/AADB2C) for Resource Manager, which (since it can't be regenerated) will want to be sourced from a different path.

I was working on the assumption that we'd discover any Sources of API Definitions within this path, which means we don't need to change the Data API to serve a new set of data (e.g. if we needed to handcraft something for MSGraph) - rather than having the directories hard-coded (which feels like it'd be hard to discover, since the `directory` argument here is the path to `./api-definitions`)?

I guess we could do the same thing by parsing a configuration file within `./api-definitions` to define which directory houses what type of data - but having the Data API know both the path to the `api-definitions` folder (provided), and then the sub-directories within it as a part of the Data API feels more magical to me?

File: tools/data-api/internal/repositories/services.go
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:405]: I think this is already supported under the `commonTypes` endpoint - we build a unique list of all the Constant definitions across all resources/versions/services, or is this different?

https://github.com/hashicorp/pandora/blob/main/tools/data-api/internal/endpoints/v1/common_types.go#L26-L53

File: tools/data-api/internal/repositories/services.go
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:190]: Could you elaborate on the potential downstream issues that could occur if we didn't duplicate the validation from saving the API definitions? I currently can't wrap my head around what we would be guarding against (except for issues in handwritten services). I feel like duplication of the validation is unnecessary bloating in the mapping for the data API code.

Furthermore I still don't understand why we have an intermediate mapping stage in the data API, it currently goes:

(for reference -> represents a mapping from a struct to a struct in the crude "diagrams" below)

**_data api models (our JSON files) -> ServiceDetail struct in the Data API -> v1 endpoints_**

The reasoning given the last time this was discussed is to maintain backward compatibility and to allow the exposure of different versioned endpoints (e.g. v2 endpoints?).

But I as far as I can tell we have the freedom to map the data api models into as many endpoints as we wish so omitting the intermediate `ServiceDetails` would achieve the same thing?

**_data api models (our JSON files) -> v1 endpoints_**

File: tools/data-api/internal/repositories/services.go
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:34]: It would help if you could provide a complete example of the directory structure you're expecting in `api-definitions` because now I'm unsure we have the same understanding.

Comment (tombuildsstuff) [reply tools/data-api/internal/repositories/services.go:34]: Sure, so I'm working on the assumption that `./api-definitions` looks like this:

```
 $ tree
.
├── README.md
├── handwritten-resource-manager
│   ├── AADB2C
│   │   ├── 2020-01-01
│   │   │   ├── ApiVersionDefinition.json
│   │   │   └── Example
│   │   │       ├── Constant-Example.json
│   │   │       ├── Model-ExampleResource.json
│   │   │       ├── Operation-Get.json
│   │   │       └── ResourceId-Example.json
│   │   └── ServiceDefinition.json
│   └── metadata.json
├── microsoft-graph
│   ├── Applications
│   │   ├── V1
│   │   │   ├── ApiVersionDefinition.json
│   │   │   └── Example
│   │   │       ├── Constant-Example.json
│   │   │       ├── Model-ExampleResource.json
│   │   │       ├── Operation-Get.json
│   │   │       └── ResourceId-Example.json
│   │   └── ServiceDefinition.json
│   └── metadata.json
└── resource-manager
    ├── Advisor
    │   ├── 2020-01-01
    │   │   ├── ApiVersionDefinition.json
    │   │   └── Example
    │   │       ├── Constant-Example.json
    │   │       ├── Model-ExampleResource.json
    │   │       ├── Operation-Get.json
    │   │       └── ResourceId-Example.json
    │   ├── 2022-10-01
    │   │   ├── ApiVersionDefinition.json
    │   │   └── Example
    │   │       ├── Constant-Example.json
    │   │       ├── Model-ExampleResource.json
    │   │       ├── Operation-Get.json
    │   │       └── ResourceId-Example.json
    │   └── ServiceDefinition.json
    └── metadata.json
```

That is, multiple directories can contain the API Definitions for Resource Manager:

* `./api-definitions/handwritten-resource-manager` - contains the handwritten API Definitions.
* `./api-definitions/resource-manager` - contains the generated API Definitions.

And (at present) only one directory for the API Definitions for Microsoft Graph:

* `./api-definitions/microsoft-graph` - contains the generated API Definitions.

(Aside: @manicminer wants to update how MS Graph outputs it's data, such that it'll be using a single `microsoft-graph` endpoint with versioning done ala ARM (e.g. `./api-definitions/microsoft-graph/Applications/(V1|Beta)` - but that's not something we need to handle now / can be dealt with in the future).

Each of the sub-directories contains a `metadata.json` file which we can use to determine which Service Type this is associated with:

```json
{
	"dataSource": "AzureResourceManager",
	"sourceInformation": "Azure/azure-rest-api-specs",
	"gitRevision": "13ef12f5de907f32bb6cac13123824c48c911411"
}
```

Which we can use to build up a canonical set of Data for each Service Type - so that we can serve this via a single endpoint.

It's worth noting that when building up the list of supported Services, we'll want to validate that each Service exists only in a single set of data (that is, `Storage` can exist in both `handwritten-resource-manager` and `resource-manager` - but not both). In the event we need to hand-define a Service for some reason, using a more specific name for the handwritten set of data shouldn't be a big deal.

File: tools/data-api/internal/repositories/services.go
Comment (tombuildsstuff) [reply tools/data-api/internal/repositories/services.go:232]: > Furthermore I still don't understand why we have an intermediate mapping stage in the data API

We recently discussed outputting the Tests into `*.hcl` files - which (already) means the data represented on disk isn't going to be 1:1 with the models we're wanting to return. For example, it might seem like a safe assumption that we can rely on the existence of tests in the structure `tests/{name}.hcl` - however the test setup also supports custom tests with multiple stages (see below) - as such the files persisted on disk are going to want to contain the path to the HCL file for that test (stage):

```json
{
  "tests": {
    "basic": "tests/basic.hcl",
    "complete": "tests/complete.hcl",
    "requiresImport": "tests/requiresImport.hcl",
    "other": {
      "first": [
        "tests/first-01.hcl",
        "tests/first-02.hcl"
      ]
    }
  }
}
```

However when returning the `TerraformResourceTestsDefinition` struct from the repository we shouldn't expose the file names (since these are only an implementation detail) - so mapping this data allows us to instead return the test configurations instead (so that other parts of the codebase don't rely on the file path) - giving a clear separation of concerns here.

Note that whilst we aren't using the custom test infrastructure today, it's hooked up through the codebase with the intention to be used for more complex resources, so we should continue to support that so we can utilize it as needed.

> The reasoning given the last time this was discussed is to maintain backward compatibility and to allow the exposure of different versioned endpoints (e.g. v2 endpoints?).

In addition to above - having the `repositories` package define the model which gets returned means that things are considerably more flexible, since we can both:

1) Change the backing store (e.g. to use a database [like we discussed previously], using a different set of data) - without rewriting the rest of the codebase.

2) Change the shape of the data in a more gradual fashion (e.g. splitting `Reference` into `ConstantReference` and `ModelReference` - and gradually updating usages to account for that)

We are still going to need to map from the types used in the `repository` package onto the `v1` endpoints, since as you've mentioned these are versioned endpoints that other tooling will be relying on - so doing this makes it easier to provide compatibility guarantees there.

Which I guess follows into:

> Could you elaborate on the potential downstream issues that could occur if we didn't duplicate the validation from saving the API definitions? 

At this point in time the Importer(s) are writing the API Definitions to disk (to keep this the same as the V1 Data API) - however we've talked about having the Data API be the only tool which manages the files in the future - meaning that it's likely we could update this to use a POST endpoint in the future (for example).

As such having this validation this validation:

1. Continues providing compatibility guarantees at each stage of Pandora's pipeline (since the downstream consumers can rely on the shape/payload) - which makes merging far less risky and means we have faster feedback (since if there's an error in the pipeline we can fix it and move forward). This also means that data issues can be caught in the part of the project where the issue exists - so (for example) the Generators can rely on the data from the Data API.
2. By being located in the Repository, which is loaded when the API is launched - and not relegated to a unit test, this validation becomes a mandatory code path - meaning that data issues are far more likely to be caught.
3. This also allows the `watch` command to provide faster feedback too - similar to 1 and 2 above.
4. Allows the Data API to lean on this when inserting/updating/deleting data in the future, when the Importer POSTs data to the Data API.

It's worth noting that Pandora initially didn't have these checks, which meant that diagnosing problems took far longer, which is why these got added to V1 (and in the other places) since we were losing time to it - from memory this has previously caught:

* Object Definitions being incorrect
* The Data API not correctly serving Models which were used as Discriminated Types, meaning these were output as regular models by mistake.
* The Data API not correctly serving Constants used in the (Request) Options object.

As such whilst this does increase the complexity here (since we're going to need to maintain the list of Constants persisted to disk and transform those here too), doing so allows us to lean on these errors rather than requiring us to jump into the debugger to track down the source of data issues, providing for far faster feedback (which also gets run in a unit test at PR time, when new data is generated).

Finally, longer term the Data API is likely to become the sole maintainer of this data, meaning that we'd likely want the same logic to validate the data that gets persisted to disk (handling generated data), in addition to when the data is loaded from disk (handling both generated and handwritten data) - meaning that this will become centralised - but that's one for the future.

**Review by tombuildsstuff — APPROVED** [review]

**Pull request merged**
